prometheus-to-sd/translator/translator.go

```diff
@@ -42,28 +43,59 @@ var supportedMetricTypes = map[dto.MetricType]bool{
 
 const falseValueEpsilon = 0.001
 
-// TranslatePrometheusToStackdriver translates metrics in Prometheus format to Stackdriver format.
-func TranslatePrometheusToStackdriver(config *config.CommonConfig,
-	whitelisted []string,
-	metrics map[string]*dto.MetricFamily,
-	cache *MetricDescriptorCache) []*v3.TimeSeries {
+type timeSeriesBuilder struct {
+	config      *config.CommonConfig
+	whitelisted []string
+	cache       *MetricDescriptorCache
+	timeSeries  map[string]*v3.TimeSeries
+}
+
+func NewTimeSeriesBuilder(config *config.CommonConfig, whitelisted []string, cache *MetricDescriptorCache) *timeSeriesBuilder {
+	return &timeSeriesBuilder{
+		config:      config,
+		whitelisted: whitelisted,
+		cache:       cache,
+		timeSeries:  make(map[string]*v3.TimeSeries),
+	}
+}
+
+// Builds a new TimeSeries array and restarts the internal state.
+func (t *timeSeriesBuilder) Build() []*v3.TimeSeries {
+	s := make([]*v3.TimeSeries, 0, len(t.timeSeries))
+	for _, v := range t.timeSeries {
+		s = append(s, v)
+	}
+	t.timeSeries = make(map[string]*v3.TimeSeries)
+	return s
+}
 
+// Update translates metrics in Prometheus format to Stackdriver format and updates the current batch.
+func (t *timeSeriesBuilder) Update(metrics map[string]*dto.MetricFamily) {
 	startTime := getStartTime(metrics)
-	metrics = filterWhitelisted(metrics, whitelisted)
+	metrics = filterWhitelisted(metrics, t.whitelisted)
 
-	var ts []*v3.TimeSeries
 	for name, metric := range metrics {
-		if cache.IsMetricBroken(name) {
+		if t.cache.IsMetricBroken(name) {
 			continue
 		}
-		t, err := translateFamily(config, metric, startTime, cache)
+		f, err := translateFamily(t.config, metric, startTime, t.cache)
 		if err != nil {
 			glog.Warningf("Error while processing metric %s: %v", name, err)
 		} else {
-			ts = append(ts, t...)
+			for _, v := range f {
```

Comment:
From what I have understood you are trying to collect latest values for all possible combinations of metric labels. Not sure that it worth doing it. First it's very rare pattern when someone resets prometheus collector, so usually the last scrape will give you the same result as you are trying to achieve here.

Reply:
Ok, sounds reasonable, PTAL.